cf-net/cf-net.c

```diff
@@ -0,0 +1,507 @@
+    }
+    else if (S_ISLNK(st_mode))
+    {
+        nprint("'%s' is a symoblic link.\n", file);
+    }
+    else
+    {
+        nprint("'%s' has an unrecognized st_mode.\n", file);
+    }
+}
+
+int CFNetStat(char **args)
+{
+    char *server = args[0];
+    char *file = args[1];
+    AgentConnection *conn = CFNetOpenConnection(server);
+    if (conn == NULL)
+    {
+        return -1;
+    }
+    bool encrypt = true;
+    struct stat sb;
+    int r = cf_remote_stat(conn, encrypt, file, &sb, "file");
+    if ( r!= 0 )
+    {
+        nprint("Could not stat: '%s'.\n", file);
+        // nprint("(Check cf-serverd output for more info)\n");
+    }
+    else
+    {
+        Log(LOG_LEVEL_INFO, "Detailed stat output:\n"
+                            "mode  = %jo, \tsize = %jd,\n"
+                            "uid   = %ju, \tgid = %ju,\n"
+                            "atime = %jd, \tmtime = %jd\n",
+            (uintmax_t)sb.st_mode,(intmax_t)sb.st_size,
+            (uintmax_t)sb.st_uid, (uintmax_t)sb.st_gid,
+            (intmax_t)sb.st_atime,(intmax_t) sb.st_mtime);
+        CFNetStatPrint(file, sb.st_mode);
+    }
+    DeleteAgentConn(conn);
+    return 0;
+}
+
+int JustConnect(char *server, char* port)
+{
+    char txtaddr[CF_MAX_IP_LEN] = "";
+    return SocketConnect(server, port, 100, true,
+                         txtaddr, sizeof(txtaddr));
+}
+
+int CFNetMulti(char *server)
+{
+    time_t start;
+    time(&start);
+
+    int ret = 0;
+    int attempts = 0;
+
+    nprint("Connecting repeatedly to '%s' without handshakes.\n",server);
+    attempts = 0;
+
+    char *buf = xstrdup(server);
+    char *host, *port;
+    ParseHostPort(buf, &host, &port);
+    if (port == NULL)
+    {
+        port = CFENGINE_PORT_STR;
+    }
+
+    ret = 0;
+    while(ret != -1)
+    {
+        ret = JustConnect(host, port);
+        attempts++;
+    }
+
+    free(buf);
+    nprint("Server unavailable after %i attempts.\n", attempts);
+
+    time_t stop;
+    time(&stop);
+    double seconds = difftime(stop,start);
+    nprint("%s(): '%s' unavailable after %.f seconds (%i attempts).\n", __FUNCTION__, server, seconds, attempts);
+    return 0;
+}
+
+int CFNetMultiTLS(char *server)
+{
+    time_t start;
+    time(&start);
+
+    nprint("Multiple handshakes to '%s'.\n", server);
+    cfnet_silent = true;
+    int ret = 0;
+    int attempts = 0;
+    while(ret == 0)
+    {
+        ret = CFNetConnectSingle(server);
+        ++attempts;
+    }
+    cfnet_silent = false;
+    time_t stop;
+    time(&stop);
+
+    double seconds = difftime(stop,start);
+    nprint("%s(): '%s' unavailable after %.f seconds (%i attempts).\n", __FUNCTION__, server, seconds, attempts);
+    CFNetMulti(server);
+    return 0;
+}
+
+#define CFNetIfMatchRun(cmd_string, comparison, functioncall)\
+    if (strcmp(cmd_string, comparison) == 0)\
+    {\
+        return functioncall;\
+    }\
+
+// Macro returns on match(!)
+int CFNetCommandSwitch(char *cmd, char **args)
```

Comment:
I might have misunderstood this one. What does "return 1" mean, in this context? Doesn't it just mean success?